relay-server/src/actors/store.rs

```diff
@@ -58,33 +58,143 @@ fn make_distinct_id(s: &str) -> Uuid {
+        }
     }
+}
 
-    let mut client_config = ClientConfig::new();
+/// This object containes the Kafka producer variants for single and sharded configurations.
+pub enum Producer {
+    Single {
+        topic_name: String,
+        producer: Arc<ThreadedProducer>,
+    },
+    Sharded(Sharded),
+}
 
-    for config_p in kafka_config {
-        client_config.set(config_p.name.as_str(), config_p.value.as_str());
+/// Sharded producer configuration.
+pub struct Sharded {
+    /// The maximum number of shards for this producer.
+    shards: u64,
+    /// The actual Kafka producer assigned to the range of logical shards, where the `u64` in the map is
+    /// the inclusive beginning of the range.
+    producers: BTreeMap<u64, (String, Arc<ThreadedProducer>)>,
+}
+
+impl Sharded {
+    /// Returns topic name and the Kafka producer based on the provided sharding key.
+    pub fn get_producer(&self, sharding_key: u64) -> (&str, Arc<ThreadedProducer>) {
+        let mut hasher = FnvHasher::default();
+        std::hash::Hash::hash(&sharding_key, &mut hasher);
+        let org_id_hash = hasher.finish();
+        let shard = org_id_hash % self.shards;
+
+        // Safety:
+        // We always have at least 1 shard so it's safe to unwrap this value.
+        let (topic_name, producer) = self
+            .producers
+            .iter()
+            .take_while(|(k, _)| *k <= &shard)
+            .last()
+            .map(|(_, v)| v)
+            .expect("At least one shard is defined");
+
+        (topic_name, producer.clone())
     }
+}
 
-    let producer = Arc::new(
-        client_config
-            .create_with_context(CaptureErrorContext)
-            .context(ServerErrorKind::KafkaError)?,
-    );
+impl Producer {
+    /// Sends the message to Kafka using correct producer for the current topic.
+    fn send(&self, organization_id: u64, message: KafkaMessage) -> Result<(), StoreError> {
+        let serialized = message.serialize()?;
+        metric!(
+            histogram(RelayHistograms::KafkaMessageSize) = serialized.len() as u64,
+            variant = message.variant()
+        );
+        let key = message.key();
+
+        match self {
+            Self::Single {
+                topic_name,
+                producer,
+            } => {
+                let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
+                producer
+                    .send(record)
+                    .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;
+            }
+
+            Self::Sharded(sharded) => {
+                let (topic_name, producer) = sharded.get_producer(organization_id);
+                let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
+                producer
+                    .send(record)
+                    .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;
+            }
+        }
```

Comment:
There's code duplication in here, might be nice to only use the `match` statement to obtain the topic name & producer:

```suggestion
        let (topic_name, producer) = match self {
            Producer::Single {
                topic_name,
                producer,
            } => (topic_name.as_str(), producer),
            Producer::Sharded(sharded) => sharded.get_producer(organization_id),
        };

        let record = BaseRecord::to(topic_name).key(&key).payload(&serialized);
        producer
            .send(record)
            .map_err(|(kafka_error, _message)| StoreError::SendFailed(kafka_error))?;

```